File: continuous/distributions/rayleigh.py
import math
import scipy.stats


class RAYLEIGH:
    """
    Rayleigh distribution
    https://en.wikipedia.org/wiki/Rayleigh_distribution
    """

    def __init__(self, measurements):
        self.parameters = self.get_parameters(measurements)
        self.gamma = self.parameters["gamma"]
        self.sigma = self.parameters["sigma"]

    def cdf(self, x: float) -> float:
        """
        Cumulative distribution function
        Calculated using the definition of the function
        Alternative: quadrature integration method
        """
        z = lambda t: (t - self.gamma) / self.sigma
        return 1 - math.exp(-0.5 * (z(x) ** 2))

    def pdf(self, x: float) -> float:
        """
        Probability density function
        Calculated using definition of the function in the documentation
        """
        z = lambda t: (t - self.gamma) / self.sigma
        return z(x) * math.exp(-0.5 * (z(x) ** 2)) / self.sigma

    def get_num_parameters(self) -> int:
        """
        Number of parameters of the distribution
        """
        return len(self.parameters)

    def parameter_restrictions(self) -> bool:
        """
        Check parameters restrictions
        """
        v1 = self.sigma > 0
        return v1

    def get_parameters(self, measurements) -> dict[str, float | int]:
        """
        Calculate proper parameters of the distribution from sample measurements.
        The parameters are calculated by solving the equations of the measures expected
        for this distribution.The number of equations to consider is equal to the number
        of parameters.

        Parameters
        ==========
        measurements: MEASUREMESTS
            attributes: mean, std, variance, skewness, kurtosis, median, mode, min, max, length, num_bins, data

        Returns
        =======
        parameters : dict
            {"gamma": * , "sigma": * }
        """
        ## Scipy Rayleigh estimation
        # scipy_params = scipy.stats.rayleigh.fit(measurements.data)
        # parameters = {"gamma": scipy_params[0], "sigma": scipy_params[1]}

        ## Location and sigma solve system
        sigma = math.sqrt(measurements.variance * 2 / (4 - math.pi))
        gamma = measurements.mean - sigma * math.sqrt(math.pi / 2)

        parameters = {"gamma": gamma, "sigma": sigma}
        return parameters


if __name__ == "__main__":
    # Import function to get measurements
    import sys

    sys.path.append("../measurements")
    from measurements_continuous import MEASUREMENTS_CONTINUOUS

    # Import function to get measurements
    def get_data(direction):
        sample_distribution_file = open(direction, "r")
        data = [float(x.replace(",", ".")) for x in sample_distribution_file.read().splitlines()]
        return data

    # Distribution class
    path = "../data/data_rayleigh.txt"
    data = get_data(path)
    measurements = MEASUREMENTS_CONTINUOUS(data)
    distribution = RAYLEIGH(measurements)

    print(distribution.get_parameters(measurements))
    print(distribution.cdf(measurements.mean))
    print(distribution.pdf(measurements.mean))

File: script-files/measures/outputsContinuous/rayleigh.js
jStat = require("../node_modules/jstat");

dists = {
    rayleigh: {
        measurements: {
            nonCentralMoments: function (k, gamma, sigma) {
                return undefined;
            },
            centralMoments: function (k, gamma, sigma) {
                return undefined;
            },
            stats: {
                mean: function (gamma, sigma) {
                    return sigma * Math.sqrt(Math.PI / 2) + gamma;
                },
                variance: function (gamma, sigma) {
                    return sigma * sigma * (2 - Math.PI / 2);
                },
                standardDeviation: function (gamma, sigma) {
                    return Math.sqrt(this.variance(gamma, sigma));
                },
                skewness: function (gamma, sigma) {
                    return 0.6311;
                },
                kurtosis: function (gamma, sigma) {
                    return (24 * Math.PI - 6 * Math.PI * Math.PI - 16) / ((4 - Math.PI) * (4 - Math.PI)) + 3;
                },
                median: function (gamma, sigma) {
                    return dists.rayleigh.measurements.ppf(0.5, gamma, sigma);
                },
                mode: function (gamma, sigma) {
                    return gamma + sigma;
                },
            },
        }
    }
}
console.log(dists.rayleigh.measurements.stats.mean(10, 2))
console.log(dists.rayleigh.measurements.stats.variance(10, 2))
console.log(dists.rayleigh.measurements.stats.standardDeviation(10, 2))
console.log(dists.rayleigh.measurements.stats.skewness(10, 2))
console.log(dists.rayleigh.measurements.stats.kurtosis(10, 2))
// console.log(dists.rayleigh.measurements.stats.median(10, 2))
console.log(dists.rayleigh.measurements.stats.mode(10, 2))

// mean_value: 12.506628274631
// variance_value: 1.7168146928204138
// standard_deviation_value: 1.3102727551240672
// skewness_value: 0.6311
// kurtosis_value: 3.245089300687639
// median_value: 12.35482004503095
// mode_value: 12